packages/hash/frontend/src/lib/subgraph.ts

```diff
@@ -194,6 +194,91 @@ export const getEntity = (
   return vertex.inner;
 };
 
+/**
+ * Gets an `Entity` by its `entityId` from within the vertices of the subgraph. Throws an error if
+ * the entity couldn't be found.
+ *
+ * @param params.subgraph
+ * @param params.entityId
+ * @throws if the vertex isn't an `EntityVertex`
+ * @throws if a vertex with `entityId` doesn't exist
+ *
+ * @todo - version is required to identify a specific instance of an entity
+ *   https://app.asana.com/0/1202805690238892/1203214689883091/f
+ */
+export const mustGetEntity = (params: {
+  subgraph: Subgraph;
+  entityId: string;
+}): Entity => {
+  const { entityId, subgraph } = params;
+
+  const entity = getEntity(subgraph, entityId);
+
+  if (!entity) {
+    throw new Error(`Vertex with entityId "${entityId}" does not exist.`);
+  }
+
+  return entity;
+};
+
+/**
+ * Gets the outgoing links of an entity.
+ *
+ * @param params.subgraph
+ * @param params.entityId - the entity id of the source entity.
+ * @param params.linkTypeId (optional) - the id of the link type
+ *
+ * @todo - version is required to identify a specific instance of an entity
+ *   https://app.asana.com/0/1202805690238892/1203214689883091/f
+ */
+export const getOutgoingLinksOfEntity = (params: {
+  entityId: string;
+  subgraph: Subgraph;
+  linkTypeId?: string;
+}): LinkVertex[] => {
+  const { entityId, subgraph, linkTypeId } = params;
+
+  const outgoingLinks = subgraph.edges[entityId]!.filter(
+    ({ edgeKind }) => edgeKind === "HAS_LINK",
+  ).map(({ destination }) => subgraph.vertices[destination] as LinkVertex);
+
+  return linkTypeId
+    ? outgoingLinks.filter(({ inner }) => inner.inner.linkTypeId === linkTypeId)
+    : outgoingLinks;
+};
+
+/**
+ * Gets the incoming links of an entity.
+ *
+ * @param params.subgraph
+ * @param params.entityId - the entity id of the source entity.
+ * @param params.linkTypeId (optional) - the id of the link type
+ *
+ * @todo - version is required to identify a specific instance of an entity
+ *   https://app.asana.com/0/1202805690238892/1203214689883091/f
+ */
+export const getIncomingLinksOfEntity = (params: {
+  entityId: string;
+  subgraph: Subgraph;
+  linkTypeId?: string;
+}): LinkVertex[] => {
+  const { entityId, subgraph, linkTypeId } = params;
+
+  /** @todo: return the incoming links of an entity in a more efficient representation */
```

Comment:
@Alfred-Mountfield is there an existing tag I can link for this? (if this is something we're planning to address)

Reply:
I don't understand your TODO. What do you mean by more efficient?

We do expect the shape of this to change though, yes.

Reply:
Currently getting the outgoing links of an entity is more efficient than getting the incoming links, because the `vertices` map supports a direct lookup for the outgoing links of a entity vertex directly using the `HAS_LINK` edge, but there is no equivalent functionality for getting incoming links of an entity. I therefore have to iterate through all the values in the map to extract the incoming links, rather than performing a direct lookup.

I'm expecting we'll need to come up for a solution to this, perhaps by having both a `HAS_OUTGOING_LINK` and `HAS_INCOMING_LINK` edges, or something else?

Reply:
> I'm expecting we'll need to come up for a solution to this

Gotcha, that's what I was unclear on. I'm not sure that's a fair expectation as such, I think it's up to debate at the moment and we're going to need to figure out what we're optimising for. I wouldn't be against us storing a reverse mapping but it will continue to blow up the size of the `Subgraph` object, so it's almost certainly something we'll want to be configurable if we even do support it.

I've raised this before (as in the topic of storing a reverse index) and at the time it was deemed fine to just loop through the object. I did try and create the `Subgraph` object specifically so that we would be able to patch in a reverse index though. 

@CiaranMn, should we consider adding this now, or should we wait and see if performance proves to be a pain. I think we're going to continue using this `lib/subgraph.ts` file as the interface to the `Subgraph` (which might sort of behave as a std-lib for it which we can publish to the BP) so it maybe won't be a big surface area to actually update if we do want to add this in later.

Reply:
> I did try and create the Subgraph object specifically so that we would be able to patch in a reverse index though.

What will its name be when we have it? Is it worth renaming `HAS_LINK` to `HAS_OUTGOING_LINK` (a) in the expectation of adding incoming, (b) so it's obvious from its name which it is? appreciate `destination` suggests it's outgoing

I don't think we need either for performance reasons right now, (i.e. incoming or outgoing lookup vs looping) but we probably will at some point.

Reply:
`HAS_LINK` is probably going to be removed anyway as we're rewriting that portion of the subgraph object. But yes, will consult better namings for ingoing/outgoing edges